cdc/verification/checker.go

```diff
@@ -0,0 +1,220 @@
+			continue
+		}
+		tables = append(tables, t)
+	}
+
+	return tables, nil
+}
+
+type columnInfo struct {
+	Field   string
+	Type    string
+	Null    string
+	Key     string
+	Default *string
+	Extra   string
+}
+
+func (c *checker) getColumns(ctx context.Context, tableName string) ([]columnInfo, error) {
+	rows, err := c.db.QueryContext(ctx, fmt.Sprintf("SHOW COLUMNS FROM %s", tableName))
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			log.Error("getColumns close rows failed", zap.Error(err))
+		}
+	}()
+
+	result := []columnInfo{}
+	for rows.Next() {
+		var t columnInfo
+		if err = rows.Scan(&t.Field, &t.Type, &t.Null, &t.Key, &t.Default, &t.Extra); err != nil {
+			return result, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+		}
+		result = append(result, t)
+	}
+
+	return result, nil
+}
+
+func (c *checker) doChecksum(ctx context.Context, columns []columnInfo, databaseName, tableName string) (string, error) {
+	var columnNames, isNull []string
+	for _, item := range columns {
+		columnNames = append(columnNames, item.Field)
+
+		t := fmt.Sprintf("ISNULL(%s)", item.Field)
+		isNull = append(isNull, t)
+	}
+
+	a := strings.Join(columnNames, ",")
+	b := strings.Join(isNull, ",")
+
+	// ref: https://www.percona.com/doc/percona-toolkit/LATEST/pt-table-checksum.html
+	// TODO: hash function as a option
+	concat := fmt.Sprintf("CONCAT_WS(',', %s, %s)", a, b)
+	// nolint:gosec
+	query := fmt.Sprintf("SELECT BIT_XOR(CAST(crc32(%s) AS UNSIGNED)) AS checksum FROM %s", concat, tableName)
+
+	log.Debug("do checkSum",
+		zap.String("db", databaseName),
+		zap.String("table", tableName),
+		zap.String("query", query))
+	var checkSum string
+	err := c.db.QueryRowContext(ctx, query).Scan(&checkSum)
+	return checkSum, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+}
+
+// TODO: use ADMIN CHECKSUM TABLE for tidb if needed
+var compareCheckSum = func(ctx context.Context, upstreamChecker, downstreamChecker checkSumChecker, f *filter.Filter) (bool, error) {
+	dbs, err := upstreamChecker.getAllDBs(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	for _, db := range dbs {
+		sourceCheckSum, err := upstreamChecker.getCheckSum(ctx, db, f)
+		if err != nil {
+			return false, err
+		}
+
+		sinkCheckSum, err := downstreamChecker.getCheckSum(ctx, db, f)
+		if err != nil {
+			return false, err
+		}
+
+		if len(sourceCheckSum) != len(sinkCheckSum) {
+			log.Warn("source and sink have different checker size",
+				zap.Any("source", sourceCheckSum),
+				zap.Any("sink", sinkCheckSum))
+		}
+
+		for k, v := range sourceCheckSum {
+			target, ok := sinkCheckSum[k]
+			if !ok {
+				log.Warn("cannot find checker at sink, it may eligible to replicate",
+					zap.String("sourceChecker", v),
+					zap.String("tableName", k))
+				continue
+			}
+			if v != target {
+				log.Error("checker mismatch", zap.String("source", v), zap.String("sink", target))
+				return false, nil
```

Comment:
how about always check all tables, this would stop on the first unmatched table.

Reply:
for debug ?

Reply:
yes, if necessary.